### PR TITLE
Fix wrong console cursour position with long input text

### LINKED ISSUE #1133: Wrong console cursor position with long input text
**Version**
3.6.0

**Description**
When the input text in console exceeds a certain length, the cursor seems to be placed at half to one character before the position it was supposed to be.
image
image

The cursor is at the end of line in both pictures

### PR DESCRIPTION
**Pull request by Sardelka9515**
Also changed cursor from "|" symbol to rectangle as it's less likely to be affected by screen resolution - rectangle takes centre as drawing position.

### CODE REVIEW

**Comment by Sardelka9515** [comment]
Made the commit message more descriptive

File: source/core/Console.cs [hunk 1]
```diff
@@ -721,13 +727,30 @@ static unsafe void DisableControlsThisFrame()
 				NativeFunc.InvokeInternal(0x351220255D64C155 /*ENABLE_CONTROL_ACTION*/, 0, i, 0);
 		}
 
-		static unsafe float GetTextLength(string text)
+		static unsafe float GetTextLength(char* str, int count, float marginLength)
 		{
+			var calculated = count;
+			if (calculated > 50) { calculated = 50; }
```
Comment (kagikn) [inline source/core/Console.cs:733]: I wonder where this `50` comes from.

Comment (Sardelka9515) [reply source/core/Console.cs:733]: Tried with 99, but the issue persists. So i just picked a random moderate number

**Comment by kagikn** [comment]
I thought text margin would be different if `ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME` is called multiple times for one text label drawing, but looks like it's not the case. For this reason, I don't think there's no need to call `GetTextLength` recursively.
In this snippet, `width2` and `width3` will match and the feed will show `Text widths are the same!`.
```
Function.Call(Hash.SET_TEXT_FONT, 0);
Function.Call(Hash.SET_TEXT_SCALE, 0.35f, 0.35f);
Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "ABCDEF");
var width1 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

Function.Call(Hash.SET_TEXT_FONT, 0);
Function.Call(Hash.SET_TEXT_SCALE, 0.35f, 0.35f);
Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "ABC");
Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "DEF");
var width2 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

if (width1 == width2)
{
  GTA.UI.Notification.Show("Text widths are the same!");
}
else
{
  GTA.UI.Notification.Show("Text widths are not the same!");
}
```
Also note that `END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT` returns very small value like `0.001` if no characters are pushed, which is different from this snippet:
```
float GetTextMargin()
{
  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.35f, 0.35f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "A"); // any letter is allowed, even if font file does not contain the glyph for that letter
  var len1 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.35f, 0.35f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "AA");
  var len2 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  return len1 - (len2 - len1); // [Margin][A] - [A] = [Margin]
}
```

**Comment by Sardelka9515** [comment]
> In this snippet, width2 and width3 will match and the feed will show Text widths are the same!.

I'm not sure i understand what you mean?
That's the case for strings less than ~99 bytes. `PushLongString`, by default already split the string into chunks and push them separately, it works well except for `END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT`, that's why I needed to begin a new text command to measure the length of the remaining string.(as suggested in #1133)
But every command takes into account the little invisible space - referred to as margin, at the start of the text label (multiple calls to `ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME` doesn't increase the margin as they're seen as single command), so i have to calculate the length of that space first and substracts it in subsequent recursive calls.

**Comment by kagikn** [comment]
Looks like that's even the case for strings more than 99 bytes, making I said wrong about text margin. Even in this example, `AboutEqual` call will return `true`.
```cs
static void TestMethod()
{
  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.30f, 0.30f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, new string('a', 99));
  var widthSplitted1 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.30f, 0.30f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, new string('a', 3));
  var widthSplitted2 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.30f, 0.30f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, new string('a', 99));
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, new string('a', 3));
  var widthNoSplitted = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  // AboutEqual guarantees the 2 float expressions have almost the same value (the small difference is caused by rounding error)
  if (AboutEqual(widthNoSplitted, (widthSplitted1 + widthSplitted2 - (margin))))
  {
    GTA.UI.Notification.Show("Text widths are the same!");
  }
  else
  {
    GTA.UI.Notification.Show("Text widths are not the same!");
  }
}

static float GetMarginLength()
{
  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.30f, 0.30f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "a");

  var width1 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  Function.Call(Hash.SET_TEXT_FONT, 0);
  Function.Call(Hash.SET_TEXT_SCALE, 0.35f, 0.35f);
  Function.Call(Hash.BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, "CELL_EMAIL_BCON");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "");
  Function.Call(Hash.ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME, "aa");

  var width12 = Function.Call<float>(Hash.END_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT, true);

  return width1 - (width12 - width1);
}

static float CalcFloatEpsilon()
{
  float eps = 1.0f;
  while (1.0f + eps / 2.0f > 1.0f)
  {
    eps /= 2.0f;
  }
  return eps;
}

public static bool AboutEqual(float x, float y)
{
  return Math.Abs(x - y) <= CalcFloatEpsilon();
}
```

**Comment by Sardelka9515** [comment]
Maybe take `maxLengthUtf8` as a parameter and make the default 99, and change it to 67 when measuring text length?

**Comment by kagikn** [comment]
I guess we could add maxLengthUtf8 as a parameter to `GetTextLength` and make the default 99. As I tested, `BEGIN_TEXT_COMMAND_GET_SCREEN_WIDTH_OF_DISPLAY_TEXT` takes up to 98 bytes (99 bytes including the terminator character) per `ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME` call.

**Pull request closed without merging**